File: vol_004/p408.cpp
#include <assert.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>

#include <iostream>
#include <algorithm>
#include <iomanip>

typedef uint32_t u32;
#define MAX_N 100000

u32 gcd(u32 x, u32 y) // euclid algorithm from wikipedia
{
    if (y > x) std::swap(x, y);
    while (y)
    {
        u32 t = y;
        y = x % y;
        x = t;
    }
    return x;
}

int main(int argc, char **argv)
{
    u32 step, mod;
    while (scanf("%u %u\n", &step, &mod) == 2)
    {
        // problem specifies these limits but fails on submission
//        assert(step and step <= MAX_N);
//        assert(mod and mod <= MAX_N);
        std::cout << std::setw(10) << std::right << step;
        std::cout << std::setw(10) << std::right << mod;
        // good sequence must have gcd(step,mod)=1 in order for each possible
        // value to be generated
        if (gcd(step, mod) == 1)
            std::cout << "    Good Choice";
        else std::cout << "    Bad Choice";
        std::cout << std::endl << std::endl;
    }
    return 0;
}

File: vol_004/p408.py
import sys

def gcd(a,b): # euclid gcd algorithm
    if a < b: a,b = b,a
    while b != 0: a, b = b, a%b
    return a

for line in sys.stdin:
    step, mod = map(int,line.split())
    # given allowed assumption fails
#    assert 1 <= step <= 100000
#    assert 1 <= mod <= 100000
    print('%10d%10d   ' % (step,mod), ('Good Choice\n' if gcd(step,mod) == 1
                                       else 'Bad Choice\n'))
